### ISSUE TITLE
Alert based downscale stuck if one action does not resolve it

### ISSUE THREAD

**Issue by baluchicken** [banzaicloud/koperator#628]
**Describe the bug**
Downscale stuck if one action does not resolve the alert. When e.g.: downscale alert fires from prometheus the operator will downscale the kafka cluster. When one action from the operator is not enough to resolve the alert, the operator just stuck and did not proceed with an another action.

**Steps to reproduce the issue:**
Create a big enough kafka cluster with 7 brokers so the example prometheus rules are starting to fire, and not one action can resolve the alert.

**Expected behaviour**
Operator should do actions in order to resolve the occurring alert as fast as possible.

**Additional context**
We are tracking the alert ID generated by prometheus inside the operator to mitigate double processing. It turns out if the alert is not resolved by  e.g.: removing one broker the prometheus will not generate a new ID. It is an odd behaviour since all the data inside the alert has been changed.

**Comment by adamantal** [comment]
Each alert in Prometheus has a fingerprint that the operator uses for tracking. The fingerprint is like a hash - if any field has changed in the alert object, then the fingerprint will also get changed.
After examination of the current alert rules, it turned out that in the description message of the current downscale alert rule the `{{ $labels.brokerId }}` field does not get substituted with its correct value (the id of the broker), since the output of the Prometheus query does not hold this piece of information.

For sake of completeness here's the Prometheus expression and its output:
```
min(kafka_server_replicamanager_partitioncount_value) by (namespace, kafka_cr) < 40
```

```
{kafka_cr="kafka",namespace="kafka"} | 31
```

By modifying the alert rule's expression to include the `brokerId` label, the downscale of multiple brokers has finished successfully in the repro scenario described above.

TLDR: no bug in the operator, the alert rules should be adjusted to include the `brokerId` information

**Comment by ashraj1** [comment]
We have tried with the suggested expressions, but still we have same issue. BY this alert one time Downscale operation works, then it ignores rest. Can you please help. 

    - alert: PartitionCountHigh
      expr: max(kafka_server_replicamanager_partitioncount)  by (namespace, kafka_cr, brokerId) > 300
      for: 1m
      labels:
        severity: alert
        alertGroup: kafka
      annotations:
        description: 'broker {{ $labels.brokerId }} has high partition count (current value is: {{ $value }})'
        summary: 'high partition count'
        brokerConfigGroup: 'default'
        command: 'upScale'
    - alert: PartitionCountLow
      expr: min(kafka_server_replicamanager_partitioncount)  by (namespace, kafka_cr, brokerId) < 250
      for: 1m
      labels:
        severity: alert
        alertGroup: kafka
      annotations:
        description: 'broker {{ $labels.brokerId }} has low partition count (current value is: {{ $value }})'
        summary: 'low partition count'
        command: 'downScale'

-------
Logs 
------
{"level":"info","ts":"2021-06-23T20:51:39.239Z","logger":"alertmanager","msg":"Stored Alert","key":"5d3e08653933590c","status":"firing","labels":"{alertGroup=\"kafka\", alertname=\"PartitionCountLow\", brokerId=\"0\", kafka_cr=\"kafka\", namespace=\"kb\", prometheus=\"kb/kafka-prometheus\", severity=\"alert\"}","annotations":"{command=\"downScale\", description=\"broker 0 has low partition count (current value is: 75)\", summary=\"low partition count\"}","processed":true}
{"level":"info","ts":"2021-06-23T20:51:39.239Z","logger":"alertmanager","msg":"Stored Alert","key":"2a02b3a1f07ca799","status":"firing","labels":"{alertGroup=\"kafka\", alertname=\"PartitionCountLow\", brokerId=\"1\", kafka_cr=\"kafka\", namespace=\"kb\", prometheus=\"kb/kafka-prometheus\", severity=\"alert\"}","annotations":"{command=\"downScale\", description=\"broker 1 has low partition count (current value is: 74)\", summary=\"low partition count\"}","processed":true}
{"level":"info","ts":"2021-06-23T20:51:39.239Z","logger":"alertmanager","msg":"Stored Alert","key":"3adf6251e1b065f6","status":"firing","labels":"{alertGroup=\"kafka\", alertname=\"PartitionCountLow\", brokerId=\"2\", kafka_cr=\"kafka\", namespace=\"kb\", prometheus=\"kb/kafka-prometheus\", severity=\"alert\"}","annotations":"{command=\"downScale\", description=\"broker 2 has low partition count (current value is: 74)\", summary=\"low partition count\"}","processed":true}

**Comment by adamantal** [comment]
Hi @ashraj1 

In order to fix the issue, modify the expression of the `PartitionCountLow` rule from:
```
  expr: min(kafka_server_replicamanager_partitioncount)  by (namespace, kafka_cr, brokerId) < 250
```
to
```
  expr: bottomk(1, kafka_server_replicamanager_partitioncount)  by (namespace, kafka_cr, brokerId) < 250
```

Adding the `brokerId` to the aggregator function of `min` is not enough, as the following series of events will occur:
1. For all brokers (there can be many), an alert will be generated (just as it can be seen in the logs you attached)
2. All these alerts are processed almost instantly by the operator, and it will react the same way to all of these alerts: removes the broker with the least partition (see code [here](https://github.com/banzaicloud/kafka-operator/blob/master/internal/alertmanager/currentalert/process.go#L313)). 
3. Since the Cruise Control's operation takes time, but the operator works almost instantly, the operator gives the same instruction for all these alerts, and believes that this single action will resolve all the alerts - thus moving these alerts to `processed` state and ignores them.
4. A single broker is removed, some partition counts may increase but continue to be under the threshold.
5. As brokers continue to have low partition count, the alert will continue to fire. Since the labels (ids of the brokers with low partition count) will not change, nor does the fingerprint of the alerts (which is based on the labels of the alert). There will be no new alerts from Prometheus's and the operator's point of view.

The solution for this issue is to change `min()` to `bottomk(1, )` which does not only contain the value, but also all the labels including the id of the broker that has this minimum value. Consequently if a single removal of broker would not resolve this issue, new alert (alert with new fingerprint because of the different `brokerId` label with the minimum) will be created.

You can see this problem as the alerts are processed all together, and not one by one - the `bottomk` enforces that we emit alert only for the least partition, broker id of which will always change, as they get removed by the operator one by one.

**Comment by ashraj1** [comment]
We have tried with the option you have suggested. Still, we are not able to do the second operation. 
[Kafka-Operator-Logs.txt](https://github.com/banzaicloud/kafka-operator/files/6730547/Kafka-Operator-Logs.txt)

**Comment by baluchicken** [comment]
@ashraj1 can you please provide your updated Prometheus alert definition?

**Comment by ashraj1** [comment]
Please find attached file for Prometheus Rules.

Ashish Patel

________________________________

**Comment by ashraj1** [comment]
[11-2-kafkacluster-prometheus.txt](https://github.com/banzaicloud/kafka-operator/files/6735033/11-2-kafkacluster-prometheus.txt)

**Comment by laukik85** [comment]
@baluchicken: Soft remainder. Can you please suggest and update?

**Comment by ashraj1** [comment]
@baluchicken @adamantal - Can you please suggest if any update on this.

**Comment by adamantal** [comment]
Hi @ashraj1,

A few things:
- I see in the logs that the upscale and the downscale alert is firing simultaneously. This is very odd. The operator also throws errors as it tries to process the downscale operation, because the Kafka custom resource should be changed to the opposite direction (simultaneously add and remove brokers) by those events.
- It seems that your Kafka cluster contains brokers with high and low partition counts at the same time. Please note that removing and adding brokers would  eventually not conclude in a balanced cluster. You should enable the CruiseControl's rebalance and healing functionality in order to achieve that. Can you check your CC configs to confirm you are using that? 
- Please note that the interval you provided is a bit strict. Let's say you have a 3-node cluster, and you configure 150 as lower and 200 as upper bound. In a high-partition count 3-node cluster (with 3*200=600 partition) an upscale event would result in a low partition count 4-node cluster (4*150=600 partition). If you have a much bigger cluster, then it's not a problem though. Can you tell me the size of the cluster?
- If `min()` had to be replaced with `bottomk(1,)`, the `max()` could be buggy too. That should be investigated on our part, I'd ask for your patience on this.
- I'd also suggest to increase the `1 min` interval in the Prometheus rule's config. I think 1 minute in a bigger cluster is not sufficient to evenly balance partition by a CruiseControl job.

Please provide your inputs on the above, I'll try to get back to you as fast as possible.

**Comment by laukik85** [comment]
Hi @adamantal,

Thanks for reply.

You should enable the CruiseControl's rebalance and healing functionality in order to achieve that. Can you check your CC configs to confirm you are using that?
**Can you please suggest which parameter I should look for to make sure rebalance and healing functionality is enabled in CC?**

 Here I am attaching cluster yaml for ref.
[10-kafkacluster.txt](https://github.com/banzaicloud/kafka-operator/files/6780615/10-kafkacluster.txt)

**Comment by stoader** [comment]
@laukik85  I suggest checking out CC documentation https://github.com/linkedin/cruise-control/wiki/Configurations and update the CC configs in your KafkaCluster CR accordingly. The CC config in KafkaCluster CR is stored under 
```
cruiseControlConfig:
  config: |
```

section.

**Comment by laukik85** [comment]
@stoader / @adamantal,

We really appreciate your help on this. We have applied appropriate configuration updates and down scaling of Kafka Brokers  worked as expected.

To match realistic scenario, we tried using network metrics to up/down scale Kafka Brokers. We used following alert expressions and we noted similar issue as of we were facing with down scale with min() using partition metrics. Can you please suggest?

rules:
- alert: BrokerOverLoaded
  expr: avg(sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m]))) by (kafka_cr, namespace) > 200
  for: 5m
  labels:
	severity: alert
	alertGroup: kafka
  annotations:
	description: 'broker {{ $labels.brokerId }} overloaded (current value is: {{ $value }})'
	summary: 'broker is overloaded'
	# If brokerConfigGroup is defined it will override scaled broker config
	brokerConfigGroup: 'default'
	command: 'upScale'
- alert: BrokerUnderLoaded
  **expr: (avg(sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m]))) by (kafka_cr, namespace) < 100)**
  for: 10m
  labels:
	severity: alert
	alertGroup: kafka
  annotations:
	description: 'broker {{ $labels.brokerId }} underloaded (current value is: {{ $value }})'
	summary: 'broker is underloaded'
	command: 'downScale'

**Comment by adamantal** [comment]
Hey @laukik85,

I'm happy that you've managed to make it work your example.

I think network based auto-scaling will not work due to limitations on the current alert processing method in the operator. The [code](https://github.com/banzaicloud/kafka-operator/blob/master/internal/alertmanager/currentalert/process.go#L313) referred above shows that the downscale will remove the broker with the least partitions regardless of any other input (like the `brokerId` label). Even though you want to scale by the network traffic, the alert generated by the low network traffic broker will be resolved in the operator by removing the broker with the least partition. Note that the two brokers (lowest network traffic and with the least partitions) may match, but are not necessarily the same. After the removal of broker, the operator considers the alert processed, so will not remove any further broker, and the alert may keep firing. 

This is actually a thing that can be improved in the operator, but it's a different issue from the one that've started off. I'll file an issue for this.


Other than that, your query
```
(avg(sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m]))) by (kafka_cr, namespace) < 100)
```
contains only the `kafka_cr` and `namespace` labels, thus will only be processed once due to the alert processing pipeline described in the [comment above](https://github.com/banzaicloud/kafka-operator/issues/628#issuecomment-867464748). You need to have a query that contains this information, like:
```
sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m])) < 100
```
but this will be triggered for all brokers with lower than 100 rate, so lowering that limit is advised (like 90). This will not result the exact same behaviour though, but approximately gives you the result. However as described above, this will not work in case the two brokers are different (which is likely the case), so currently the non partition-based scaling will not work.

**Comment by adamantal** [comment]
Hi!
With the merge of #652, I closed the improvement issue #637. Thanks for @panyuenlau for taking care of this. The next `koperator` (previously known as `kafka-operator`, we renamed this project in the meantime) release will have this improvement.
We would be pleased, if you happen to test the next release and give us some feedbacks on the autoscaling.

**Comment by laukik85** [comment]
Hi @adamantal,

Appreciate your help on this. 

We tried scaling based on network parameters and seeing saw same behavior where after removal of one broker, operator considered alert has been processed and stopped removing other candidate brokers.

I am using following rules. Can you please suggest?

rules:

alert: BrokerOverLoaded
expr: avg(sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m]))) by (kafka_cr, namespace) > 200
for: 10m
labels:
severity: alert
alertGroup: kafka
annotations:
description: 'broker {{ $labels.brokerId }} overloaded (current value is: {{ $value }})'
summary: 'broker is overloaded'
brokerConfigGroup: 'default'
command: 'upScale'
alert: BrokerUnderLoaded
expr: (avg(sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m]))) by (kafka_cr, namespace) < 100)
for: 10m
labels:
severity: alert
alertGroup: kafka
annotations:
description: 'broker {{ $labels.brokerId }} underloaded (current value is: {{ $value }})'
summary: 'broker is underloaded'
command: 'downScale'

**Comment by adamantal** [comment]
Tried to reproduce the problem, and I discovered two issues:
- your expression in the Prometheus rule is not aligned with the principles described above in my previous comment, so since the label is not provided, the operator continues to use the broker with the least number of partitions
- it seems that #637 introduced the wrong label to check, but with that fix I managed to configure autoscaling based on network metrics

**Comment by adamantal** [comment]
The downscale worked for me using the following expression instead:
```
sum by(brokerId, kafka_cr, namespace) (rate(kafka_network_requestmetrics_requests_total[10m]))) < 20
```
Note that it is important that the alert contains the `brokerId` label.

Hope that this helps.

**Comment by laukik85** [comment]
@adamantal Thanks for responding - I will try suggested and revert.

**Comment by laukik85** [comment]
@adamantal I have tried suggested and it worked as expected. Thanks.
